{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "S1_table = pd.read_csv('/lunarc/nobackup/projects/lu2024-17-19/J_rAAven/BRAVE/0_data/SortedFragments.csv')\n",
    "S4_table = pd.read_csv('/lunarc/nobackup/projects/lu2024-17-19/J_rAAven/BRAVE/0_data/library_barcodes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "S1_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "S4_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library_barcodes = pd.merge(S1_table, S4_table, on='LUTnr', how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library_barcodes.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Category', 'GeneName', 'AAstart', 'AAend', 'Structure', 'Peptide',\n",
       "       'Sequence', 'LUTnr', 'start', 'end', 'width', 'BC', 'bitScore',\n",
       "       'mismatches', 'tCount', 'mCount', 'Mode'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "library_barcodes.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library_barcodes = library_barcodes.drop(columns=[ 'Reads', 'identity','alignmentLength', 'gapOpens', 'q_start', 'q_end', 's_start', 's_end','evalue'])\n",
    "library_barcodes.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reorganize columns\n",
    "library_barcodes = library_barcodes[['Category', 'GeneName', 'LUTnr',  'BC', 'AAstart', 'AAend', 'Structure', 'Peptide',\n",
    "    'start', 'end', 'width', 'Sequence', 'bitScore',\n",
    "    'mismatches', 'tCount', 'mCount', 'RNAcount', 'Mode']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Category</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>LUTnr</th>\n",
       "      <th>BC</th>\n",
       "      <th>AAstart</th>\n",
       "      <th>AAend</th>\n",
       "      <th>Structure</th>\n",
       "      <th>Peptide</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>bitScore</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>tCount</th>\n",
       "      <th>mCount</th>\n",
       "      <th>Mode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Neurotrophin</td>\n",
       "      <td>NGF</td>\n",
       "      <td>seq_1</td>\n",
       "      <td>CATGCTATAAGGGAAGATTT</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>14aa</td>\n",
       "      <td>EPYTDSNVPEGDSV</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG</td>\n",
       "      <td>114.00</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Def</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Neurotrophin</td>\n",
       "      <td>NGF</td>\n",
       "      <td>seq_1</td>\n",
       "      <td>GTGTGTTGCCTTCATTCAGC</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>14aa</td>\n",
       "      <td>EPYTDSNVPEGDSV</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG</td>\n",
       "      <td>117.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Def</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Neurotrophin</td>\n",
       "      <td>NGF</td>\n",
       "      <td>seq_1</td>\n",
       "      <td>ACATGTTCATGTGCGCGCTG</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>14aa</td>\n",
       "      <td>EPYTDSNVPEGDSV</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG</td>\n",
       "      <td>75.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>Def</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Neurotrophin</td>\n",
       "      <td>NGF</td>\n",
       "      <td>seq_1</td>\n",
       "      <td>ATAGCAGGATGTCAGTGTGG</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>14aa</td>\n",
       "      <td>EPYTDSNVPEGDSV</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG</td>\n",
       "      <td>125.25</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>Def</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Neurotrophin</td>\n",
       "      <td>NGF</td>\n",
       "      <td>seq_1</td>\n",
       "      <td>ATGTGTTTGAAGGATGGATT</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>14aa</td>\n",
       "      <td>EPYTDSNVPEGDSV</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG</td>\n",
       "      <td>126.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>Def</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Category GeneName  LUTnr                    BC  AAstart  AAend  \\\n",
       "0  Neurotrophin      NGF  seq_1  CATGCTATAAGGGAAGATTT        0     14   \n",
       "1  Neurotrophin      NGF  seq_1  GTGTGTTGCCTTCATTCAGC        0     14   \n",
       "2  Neurotrophin      NGF  seq_1  ACATGTTCATGTGCGCGCTG        0     14   \n",
       "3  Neurotrophin      NGF  seq_1  ATAGCAGGATGTCAGTGTGG        0     14   \n",
       "4  Neurotrophin      NGF  seq_1  ATGTGTTTGAAGGATGGATT        0     14   \n",
       "\n",
       "  Structure         Peptide  start  end  width  \\\n",
       "0      14aa  EPYTDSNVPEGDSV      0   42     42   \n",
       "1      14aa  EPYTDSNVPEGDSV      0   42     42   \n",
       "2      14aa  EPYTDSNVPEGDSV      0   42     42   \n",
       "3      14aa  EPYTDSNVPEGDSV      0   42     42   \n",
       "4      14aa  EPYTDSNVPEGDSV      0   42     42   \n",
       "\n",
       "                                     Sequence  bitScore  mismatches  tCount  \\\n",
       "0  GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG    114.00         2.5       2   \n",
       "1  GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG    117.00         1.0       2   \n",
       "2  GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG     75.00         0.0       5   \n",
       "3  GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG    125.25         0.0       5   \n",
       "4  GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG    126.00         1.0       7   \n",
       "\n",
       "   mCount Mode  \n",
       "0       2  Def  \n",
       "1       2  Def  \n",
       "2       3  Def  \n",
       "3       4  Def  \n",
       "4       3  Def  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "library_barcodes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3261305, 18)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "library_barcodes = pd.read_csv('/lunarc/nobackup/projects/lu2024-17-19/J_rAAven/BRAVE/0_data/pos_library_barcodes.csv')\n",
    "library_barcodes.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BC</th>\n",
       "      <th>LUTnr</th>\n",
       "      <th>bitScore</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>tCount</th>\n",
       "      <th>mCount</th>\n",
       "      <th>Mode</th>\n",
       "      <th>Reads</th>\n",
       "      <th>identity</th>\n",
       "      <th>alignmentLength</th>\n",
       "      <th>...</th>\n",
       "      <th>Category</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>AAstart</th>\n",
       "      <th>AAend</th>\n",
       "      <th>Structure</th>\n",
       "      <th>Peptide</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTGCCTAGACTCCCAGCCTT</td>\n",
       "      <td>seq_88507</td>\n",
       "      <td>122.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Amb</td>\n",
       "      <td>CTTACGCCACCGGCAGCAGCCCCGACGTGCAGATCACCGCCAGCCT...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Theilers murine encephalomyelitis virus</td>\n",
       "      <td>TMEV-VP2</td>\n",
       "      <td>234</td>\n",
       "      <td>256</td>\n",
       "      <td>22aa</td>\n",
       "      <td>YATGSSPDVQITASLQPVNPVF</td>\n",
       "      <td>TACGCCACCGGCAGCAGCCCCGACGTGCAGATCACCGCCAGCCTGC...</td>\n",
       "      <td>702</td>\n",
       "      <td>768</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTACGTGCACGCGTCATT</td>\n",
       "      <td>seq_78536</td>\n",
       "      <td>78.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Neurotoxin</td>\n",
       "      <td>BoNT/F Hc</td>\n",
       "      <td>204</td>\n",
       "      <td>226</td>\n",
       "      <td>22aa</td>\n",
       "      <td>KVFDTELGKTEIETLYSDEPDP</td>\n",
       "      <td>AAGGTGTTCGACACCGAGCTGGGCAAGACCGAGATCGAGACCCTGT...</td>\n",
       "      <td>612</td>\n",
       "      <td>678</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTACGGACATCAGCATGC</td>\n",
       "      <td>seq_34770</td>\n",
       "      <td>76.200000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Human immunodeficiency virus type 1</td>\n",
       "      <td>HIV-gp120-TH90129_1</td>\n",
       "      <td>210</td>\n",
       "      <td>224</td>\n",
       "      <td>14aa</td>\n",
       "      <td>GPCTNVSTVQCTHG</td>\n",
       "      <td>GGCCCCTGCACCAACGTGAGCACCGTGCAGTGCACCCACGGC</td>\n",
       "      <td>630</td>\n",
       "      <td>672</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTGTGTATGGGCGTATGG</td>\n",
       "      <td>seq_21642</td>\n",
       "      <td>78.700000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Herpes simplex virus type 1</td>\n",
       "      <td>HSV-1-pUS8</td>\n",
       "      <td>393</td>\n",
       "      <td>407</td>\n",
       "      <td>14aa</td>\n",
       "      <td>GADLAEPTHPHVGA</td>\n",
       "      <td>GGCGCCGACCTGGCCGAGCCCACCCACCCCCACGTGGGCGCC</td>\n",
       "      <td>1179</td>\n",
       "      <td>1221</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GTACGCGTAGGAGTCCAC</td>\n",
       "      <td>seq_22869</td>\n",
       "      <td>78.700000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Herpes simplex virus type 1</td>\n",
       "      <td>HSV-1-pUL19</td>\n",
       "      <td>1112</td>\n",
       "      <td>1126</td>\n",
       "      <td>14aa</td>\n",
       "      <td>ATVRNPVTDMGNLP</td>\n",
       "      <td>GCCACCGTGCGGAACCCCGTGACCGACATGGGCAACCTGCCC</td>\n",
       "      <td>3336</td>\n",
       "      <td>3378</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     BC      LUTnr    bitScore  mismatches  tCount  mCount  \\\n",
       "0  GTGCCTAGACTCCCAGCCTT  seq_88507  122.000000         0.0       1       1   \n",
       "1    GTACGTGCACGCGTCATT  seq_78536   78.666667         0.0       3       3   \n",
       "2    GTACGGACATCAGCATGC  seq_34770   76.200000         0.0       5       3   \n",
       "3    GTGTGTATGGGCGTATGG  seq_21642   78.700000         0.0       2       2   \n",
       "4    GTACGCGTAGGAGTCCAC  seq_22869   78.700000         0.0       2       1   \n",
       "\n",
       "  Mode                                              Reads  identity  \\\n",
       "0  Amb  CTTACGCCACCGGCAGCAGCCCCGACGTGCAGATCACCGCCAGCCT...     100.0   \n",
       "1  Def                                                NaN       NaN   \n",
       "2  Def                                                NaN       NaN   \n",
       "3  Def                                                NaN       NaN   \n",
       "4  Def                                                NaN       NaN   \n",
       "\n",
       "   alignmentLength  ...                                 Category  \\\n",
       "0             66.0  ...  Theilers murine encephalomyelitis virus   \n",
       "1              NaN  ...                               Neurotoxin   \n",
       "2              NaN  ...      Human immunodeficiency virus type 1   \n",
       "3              NaN  ...              Herpes simplex virus type 1   \n",
       "4              NaN  ...              Herpes simplex virus type 1   \n",
       "\n",
       "              GeneName  AAstart  AAend  Structure                 Peptide  \\\n",
       "0             TMEV-VP2      234    256       22aa  YATGSSPDVQITASLQPVNPVF   \n",
       "1            BoNT/F Hc      204    226       22aa  KVFDTELGKTEIETLYSDEPDP   \n",
       "2  HIV-gp120-TH90129_1      210    224       14aa          GPCTNVSTVQCTHG   \n",
       "3           HSV-1-pUS8      393    407       14aa          GADLAEPTHPHVGA   \n",
       "4          HSV-1-pUL19     1112   1126       14aa          ATVRNPVTDMGNLP   \n",
       "\n",
       "                                            Sequence start   end  width  \n",
       "0  TACGCCACCGGCAGCAGCCCCGACGTGCAGATCACCGCCAGCCTGC...   702   768     66  \n",
       "1  AAGGTGTTCGACACCGAGCTGGGCAAGACCGAGATCGAGACCCTGT...   612   678     66  \n",
       "2         GGCCCCTGCACCAACGTGAGCACCGTGCAGTGCACCCACGGC   630   672     42  \n",
       "3         GGCGCCGACCTGGCCGAGCCCACCCACCCCCACGTGGGCGCC  1179  1221     42  \n",
       "4         GCCACCGTGCGGAACCCCGTGACCGACATGGGCAACCTGCCC  3336  3378     42  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "found_barcodes = pd.read_csv('/lunarc/nobackup/projects/lu2024-17-19/J_rAAven/BRAVE/0_data/found_barcodes/found.mRNA_30cpc_Th.csv')\n",
    "found_barcodes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# concat dataframes but only keep the columns that are in both dataframes\n",
    "combined = pd.concat([library_barcodes, found_barcodes], join='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ref_sequence_length_df(file_path:str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Reads a FASTA file containing reference sequences and returns a DataFrame with sequence ids and lengths.\n",
    "    Args:\n",
    "        file_path (str): path to the file containing the reference sequences\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: DataFrame with the reference sequence ids and lengths\n",
    "    \"\"\"\n",
    "    \n",
    "    # read in file with original sequences\n",
    "    seqs_original = list(SeqIO.parse(file_path, \"fasta\"))\n",
    "    # translate sequences to amino acids\n",
    "    seq = [str(seq_record.seq).strip() for seq_record in seqs_original]\n",
    "    # get sequence ids\n",
    "    ids = [seq_record.description for seq_record in seqs_original]\n",
    "    # get sequence categories\n",
    "    categories = [id.split(\",\")[0] for id in ids]\n",
    "    # get the GeneName\n",
    "    gene_names = [id.split(\",\")[5] for id in ids]\n",
    "    \n",
    "    # create a DataFrame with sreference sequence ids and lengths\n",
    "    ref_seq_len_df = pd.DataFrame({\"Category\": categories,\n",
    "                                    \"GeneName\": gene_names,\n",
    "                                    \"seqlength\": [len(s) for s in seq]})\n",
    "    \n",
    "    return ref_seq_len_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = get_ref_sequence_length_df('/lunarc/nobackup/projects/lu2024-17-19/J_rAAven/BRAVE/input/DNA-lib_RetrogradeTransport.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA_pscAAVlib_Prep2_1\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "test = \"DNA_pscAAVlib_Prep2_1.fastq.gz\"\n",
    "test_name=os.path.basename(test).replace('.fastq.gz', '')\n",
    "print(test_name)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
