{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2332308/1929727819.py:3: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  p005 = pd.read_csv('/lunarc/nobackup/projects/lu2024-17-19/J_rAAven/BRAVE/raav-60/p005/library_barcodes.csv')\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "p005 = pd.read_csv('/lunarc/nobackup/projects/lu2024-17-19/J_rAAven/BRAVE/raav-60/p005/library_barcodes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BC</th>\n",
       "      <th>LUTnr</th>\n",
       "      <th>bitScore</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>tCount</th>\n",
       "      <th>mCount</th>\n",
       "      <th>Mode</th>\n",
       "      <th>Reads</th>\n",
       "      <th>identity</th>\n",
       "      <th>alignmentLength</th>\n",
       "      <th>gapOpens</th>\n",
       "      <th>q_start</th>\n",
       "      <th>q_end</th>\n",
       "      <th>s_start</th>\n",
       "      <th>s_end</th>\n",
       "      <th>evalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACGAAACCTAGAAAGACGGCAACCCTC</td>\n",
       "      <td>seq_23129</td>\n",
       "      <td>41.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACGAAACCTTCAAACCTTCCAATCAGG</td>\n",
       "      <td>seq_15145</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACGAAACGAACCATCATTTAAGGCTTC</td>\n",
       "      <td>seq_1054</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACGAAACGCGCATTTACTCGAGCGCTC</td>\n",
       "      <td>seq_31374</td>\n",
       "      <td>44.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACGAAAGCCGGGCACATACCCAGGCAC</td>\n",
       "      <td>seq_1118</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            BC      LUTnr  bitScore  mismatches  tCount  \\\n",
       "0  ACGAAACCTAGAAAGACGGCAACCCTC  seq_23129      41.7         1.0       2   \n",
       "1  ACGAAACCTTCAAACCTTCCAATCAGG  seq_15145      47.3         0.0       2   \n",
       "2  ACGAAACGAACCATCATTTAAGGCTTC   seq_1054      47.3         0.0       2   \n",
       "3  ACGAAACGCGCATTTACTCGAGCGCTC  seq_31374      44.5         0.5       2   \n",
       "4  ACGAAAGCCGGGCACATACCCAGGCAC   seq_1118      47.3         0.0       2   \n",
       "\n",
       "   mCount Mode Reads  identity  alignmentLength  gapOpens  q_start  q_end  \\\n",
       "0     2.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "1     2.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "2     2.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "3     2.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "4     2.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "\n",
       "   s_start  s_end  evalue  \n",
       "0      NaN    NaN     NaN  \n",
       "1      NaN    NaN     NaN  \n",
       "2      NaN    NaN     NaN  \n",
       "3      NaN    NaN     NaN  \n",
       "4      NaN    NaN     NaN  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BC</th>\n",
       "      <th>LUTnr</th>\n",
       "      <th>bitScore</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>tCount</th>\n",
       "      <th>mCount</th>\n",
       "      <th>Mode</th>\n",
       "      <th>Reads</th>\n",
       "      <th>identity</th>\n",
       "      <th>alignmentLength</th>\n",
       "      <th>gapOpens</th>\n",
       "      <th>q_start</th>\n",
       "      <th>q_end</th>\n",
       "      <th>s_start</th>\n",
       "      <th>s_end</th>\n",
       "      <th>evalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>86017</th>\n",
       "      <td>TCGGATCGAGTATAGGCACCAGTGCAT</td>\n",
       "      <td>seq_25253</td>\n",
       "      <td>46.457576</td>\n",
       "      <td>0.0</td>\n",
       "      <td>33</td>\n",
       "      <td>33.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88397</th>\n",
       "      <td>TCGGCACCTTCCCGGATACCCTCCTTC</td>\n",
       "      <td>seq_7854</td>\n",
       "      <td>46.527586</td>\n",
       "      <td>0.0</td>\n",
       "      <td>29</td>\n",
       "      <td>29.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92573</th>\n",
       "      <td>TCGGCTCCCGTACTTGCGCACACATAT</td>\n",
       "      <td>seq_37275</td>\n",
       "      <td>47.045455</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22</td>\n",
       "      <td>22.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27291</th>\n",
       "      <td>TCGACTCCTTCCAGTACACACTCCTAC</td>\n",
       "      <td>seq_5921</td>\n",
       "      <td>47.300000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20</td>\n",
       "      <td>20.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78940</th>\n",
       "      <td>TCGCTTGAAACATTCCATCATACCATC</td>\n",
       "      <td>seq_19327</td>\n",
       "      <td>47.020000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20</td>\n",
       "      <td>20.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78206</th>\n",
       "      <td>TCGCTTCCAGTACGCAAATAAGTCTAC</td>\n",
       "      <td>seq_39413</td>\n",
       "      <td>47.300000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16</td>\n",
       "      <td>16.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14862</th>\n",
       "      <td>TCGAATTATGTAAATCTATACGGGATC</td>\n",
       "      <td>seq_5796</td>\n",
       "      <td>47.300000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16</td>\n",
       "      <td>16.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18354</th>\n",
       "      <td>TCGACACGATCCTATCCTGATTGGATT</td>\n",
       "      <td>seq_21783</td>\n",
       "      <td>47.300000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15</td>\n",
       "      <td>15.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6431</th>\n",
       "      <td>TCGAAATAAATAAGCGATTAAATACGC</td>\n",
       "      <td>seq_21914</td>\n",
       "      <td>47.300000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15</td>\n",
       "      <td>15.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72799</th>\n",
       "      <td>TCGCTACGCAGCCGTCCATCCTTCTAT</td>\n",
       "      <td>seq_33563</td>\n",
       "      <td>47.300000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15</td>\n",
       "      <td>15.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44240</th>\n",
       "      <td>TCGCAAGGCTTCATTCTTCAAAGATTT</td>\n",
       "      <td>seq_16689</td>\n",
       "      <td>47.300000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83141</th>\n",
       "      <td>TCGGAATCAACATTGCATTGTGCAATC</td>\n",
       "      <td>seq_37324</td>\n",
       "      <td>47.300000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11</td>\n",
       "      <td>11.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54490</th>\n",
       "      <td>TCGCATTGTTGCTATACAGAAATCATT</td>\n",
       "      <td>seq_20138</td>\n",
       "      <td>47.300000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11</td>\n",
       "      <td>11.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66990</th>\n",
       "      <td>TCGCCTCCCGGCATCCTTTAAAGATTC</td>\n",
       "      <td>seq_28234</td>\n",
       "      <td>47.300000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11</td>\n",
       "      <td>11.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78514</th>\n",
       "      <td>TCGCTTCCCTTAATCCAGTCAACCAAC</td>\n",
       "      <td>seq_15594</td>\n",
       "      <td>45.630000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32690</th>\n",
       "      <td>TCGATAGACGCACTCCCTTCCTGCTAC</td>\n",
       "      <td>seq_31834</td>\n",
       "      <td>47.300000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82128</th>\n",
       "      <td>TCGGAAGAAATCATCCTTCGATGATAC</td>\n",
       "      <td>seq_38793</td>\n",
       "      <td>45.260000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9831</th>\n",
       "      <td>TCGAAGGAAACCTTTGTGTGAACCTAT</td>\n",
       "      <td>seq_30743</td>\n",
       "      <td>47.300000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81851</th>\n",
       "      <td>TCGGAACGAACCCTCGTACCCTTACGC</td>\n",
       "      <td>seq_19996</td>\n",
       "      <td>46.180000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16129</th>\n",
       "      <td>TCGACACAATCCTACGCACCTATAAGT</td>\n",
       "      <td>seq_25637</td>\n",
       "      <td>47.300000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                BC      LUTnr   bitScore  mismatches  tCount  \\\n",
       "86017  TCGGATCGAGTATAGGCACCAGTGCAT  seq_25253  46.457576         0.0      33   \n",
       "88397  TCGGCACCTTCCCGGATACCCTCCTTC   seq_7854  46.527586         0.0      29   \n",
       "92573  TCGGCTCCCGTACTTGCGCACACATAT  seq_37275  47.045455         0.0      22   \n",
       "27291  TCGACTCCTTCCAGTACACACTCCTAC   seq_5921  47.300000         0.0      20   \n",
       "78940  TCGCTTGAAACATTCCATCATACCATC  seq_19327  47.020000         0.0      20   \n",
       "78206  TCGCTTCCAGTACGCAAATAAGTCTAC  seq_39413  47.300000         0.0      16   \n",
       "14862  TCGAATTATGTAAATCTATACGGGATC   seq_5796  47.300000         0.0      16   \n",
       "18354  TCGACACGATCCTATCCTGATTGGATT  seq_21783  47.300000         0.0      15   \n",
       "6431   TCGAAATAAATAAGCGATTAAATACGC  seq_21914  47.300000         0.0      15   \n",
       "72799  TCGCTACGCAGCCGTCCATCCTTCTAT  seq_33563  47.300000         0.0      15   \n",
       "44240  TCGCAAGGCTTCATTCTTCAAAGATTT  seq_16689  47.300000         0.0      14   \n",
       "83141  TCGGAATCAACATTGCATTGTGCAATC  seq_37324  47.300000         0.0      11   \n",
       "54490  TCGCATTGTTGCTATACAGAAATCATT  seq_20138  47.300000         0.0      11   \n",
       "66990  TCGCCTCCCGGCATCCTTTAAAGATTC  seq_28234  47.300000         0.0      11   \n",
       "78514  TCGCTTCCCTTAATCCAGTCAACCAAC  seq_15594  45.630000         0.0      10   \n",
       "32690  TCGATAGACGCACTCCCTTCCTGCTAC  seq_31834  47.300000         0.0      10   \n",
       "82128  TCGGAAGAAATCATCCTTCGATGATAC  seq_38793  45.260000         0.0      10   \n",
       "9831   TCGAAGGAAACCTTTGTGTGAACCTAT  seq_30743  47.300000         0.0      10   \n",
       "81851  TCGGAACGAACCCTCGTACCCTTACGC  seq_19996  46.180000         0.0      10   \n",
       "16129  TCGACACAATCCTACGCACCTATAAGT  seq_25637  47.300000         0.0      10   \n",
       "\n",
       "       mCount Mode Reads  identity  alignmentLength  gapOpens  q_start  q_end  \\\n",
       "86017    33.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "88397    29.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "92573    22.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "27291    20.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "78940    20.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "78206    16.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "14862    16.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "18354    15.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "6431     15.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "72799    15.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "44240    14.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "83141    11.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "54490    11.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "66990    11.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "78514    10.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "32690    10.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "82128    10.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "9831     10.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "81851    10.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "16129    10.0  Def   NaN       NaN              NaN       NaN      NaN    NaN   \n",
       "\n",
       "       s_start  s_end  evalue  \n",
       "86017      NaN    NaN     NaN  \n",
       "88397      NaN    NaN     NaN  \n",
       "92573      NaN    NaN     NaN  \n",
       "27291      NaN    NaN     NaN  \n",
       "78940      NaN    NaN     NaN  \n",
       "78206      NaN    NaN     NaN  \n",
       "14862      NaN    NaN     NaN  \n",
       "18354      NaN    NaN     NaN  \n",
       "6431       NaN    NaN     NaN  \n",
       "72799      NaN    NaN     NaN  \n",
       "44240      NaN    NaN     NaN  \n",
       "83141      NaN    NaN     NaN  \n",
       "54490      NaN    NaN     NaN  \n",
       "66990      NaN    NaN     NaN  \n",
       "78514      NaN    NaN     NaN  \n",
       "32690      NaN    NaN     NaN  \n",
       "82128      NaN    NaN     NaN  \n",
       "9831       NaN    NaN     NaN  \n",
       "81851      NaN    NaN     NaN  \n",
       "16129      NaN    NaN     NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p005.sort_values(by='mCount', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BC</th>\n",
       "      <th>LUTnr</th>\n",
       "      <th>bitScore</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>tCount</th>\n",
       "      <th>mCount</th>\n",
       "      <th>Mode</th>\n",
       "      <th>Reads</th>\n",
       "      <th>identity</th>\n",
       "      <th>alignmentLength</th>\n",
       "      <th>gapOpens</th>\n",
       "      <th>q_start</th>\n",
       "      <th>q_end</th>\n",
       "      <th>s_start</th>\n",
       "      <th>s_end</th>\n",
       "      <th>evalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42162315</th>\n",
       "      <td>TCGCATTCAGCGCGGACTTAAGCCAGT</td>\n",
       "      <td>seq_28082</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284262</th>\n",
       "      <td>TCGCCTCCCTGGCAGCAGTAATGGATC</td>\n",
       "      <td>seq_8625</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284280</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_24284</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284279</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_23302</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284278</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_2187</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284277</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_21349</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284276</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_19660</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284275</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_19604</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284273</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_16266</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284272</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_15016</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284271</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_15006</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284270</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_128</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284269</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_10630</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284268</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_10143</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284267</th>\n",
       "      <td>TCGCCTCCCTGGCAGCAGTAATGGATC</td>\n",
       "      <td>seq_9898</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284265</th>\n",
       "      <td>TCGCCTCCCTGGCAGCAGTAATGGATC</td>\n",
       "      <td>seq_9178</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284264</th>\n",
       "      <td>TCGCCTCCCTGGCAGCAGTAATGGATC</td>\n",
       "      <td>seq_910</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284281</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_25163</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284282</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_25280</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53284284</th>\n",
       "      <td>TCGCCTCCCTGGCAGCATCAATGCAGG</td>\n",
       "      <td>seq_26690</td>\n",
       "      <td>47.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   BC      LUTnr  bitScore  mismatches  \\\n",
       "42162315  TCGCATTCAGCGCGGACTTAAGCCAGT  seq_28082      47.3         0.0   \n",
       "53284262  TCGCCTCCCTGGCAGCAGTAATGGATC   seq_8625      47.3         0.0   \n",
       "53284280  TCGCCTCCCTGGCAGCATCAATGCAGG  seq_24284      47.3         0.0   \n",
       "53284279  TCGCCTCCCTGGCAGCATCAATGCAGG  seq_23302      47.3         0.0   \n",
       "53284278  TCGCCTCCCTGGCAGCATCAATGCAGG   seq_2187      47.3         0.0   \n",
       "53284277  TCGCCTCCCTGGCAGCATCAATGCAGG  seq_21349      47.3         0.0   \n",
       "53284276  TCGCCTCCCTGGCAGCATCAATGCAGG  seq_19660      47.3         0.0   \n",
       "53284275  TCGCCTCCCTGGCAGCATCAATGCAGG  seq_19604      47.3         0.0   \n",
       "53284273  TCGCCTCCCTGGCAGCATCAATGCAGG  seq_16266      47.3         0.0   \n",
       "53284272  TCGCCTCCCTGGCAGCATCAATGCAGG  seq_15016      47.3         0.0   \n",
       "53284271  TCGCCTCCCTGGCAGCATCAATGCAGG  seq_15006      47.3         0.0   \n",
       "53284270  TCGCCTCCCTGGCAGCATCAATGCAGG    seq_128      47.3         0.0   \n",
       "53284269  TCGCCTCCCTGGCAGCATCAATGCAGG  seq_10630      47.3         0.0   \n",
       "53284268  TCGCCTCCCTGGCAGCATCAATGCAGG  seq_10143      47.3         0.0   \n",
       "53284267  TCGCCTCCCTGGCAGCAGTAATGGATC   seq_9898      47.3         0.0   \n",
       "53284265  TCGCCTCCCTGGCAGCAGTAATGGATC   seq_9178      47.3         0.0   \n",
       "53284264  TCGCCTCCCTGGCAGCAGTAATGGATC    seq_910      47.3         0.0   \n",
       "53284281  TCGCCTCCCTGGCAGCATCAATGCAGG  seq_25163      47.3         0.0   \n",
       "53284282  TCGCCTCCCTGGCAGCATCAATGCAGG  seq_25280      47.3         0.0   \n",
       "53284284  TCGCCTCCCTGGCAGCATCAATGCAGG  seq_26690      47.3         0.0   \n",
       "\n",
       "          tCount  mCount Mode Reads  identity  alignmentLength  gapOpens  \\\n",
       "42162315       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284262       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284280       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284279       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284278       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284277       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284276       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284275       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284273       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284272       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284271       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284270       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284269       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284268       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284267       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284265       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284264       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284281       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284282       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "53284284       1     NaN  Amb   NaN       NaN              NaN       NaN   \n",
       "\n",
       "          q_start  q_end  s_start  s_end  evalue  \n",
       "42162315      NaN    NaN      NaN    NaN     NaN  \n",
       "53284262      NaN    NaN      NaN    NaN     NaN  \n",
       "53284280      NaN    NaN      NaN    NaN     NaN  \n",
       "53284279      NaN    NaN      NaN    NaN     NaN  \n",
       "53284278      NaN    NaN      NaN    NaN     NaN  \n",
       "53284277      NaN    NaN      NaN    NaN     NaN  \n",
       "53284276      NaN    NaN      NaN    NaN     NaN  \n",
       "53284275      NaN    NaN      NaN    NaN     NaN  \n",
       "53284273      NaN    NaN      NaN    NaN     NaN  \n",
       "53284272      NaN    NaN      NaN    NaN     NaN  \n",
       "53284271      NaN    NaN      NaN    NaN     NaN  \n",
       "53284270      NaN    NaN      NaN    NaN     NaN  \n",
       "53284269      NaN    NaN      NaN    NaN     NaN  \n",
       "53284268      NaN    NaN      NaN    NaN     NaN  \n",
       "53284267      NaN    NaN      NaN    NaN     NaN  \n",
       "53284265      NaN    NaN      NaN    NaN     NaN  \n",
       "53284264      NaN    NaN      NaN    NaN     NaN  \n",
       "53284281      NaN    NaN      NaN    NaN     NaN  \n",
       "53284282      NaN    NaN      NaN    NaN     NaN  \n",
       "53284284      NaN    NaN      NaN    NaN     NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p005.sort_values(by='bitScore', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BC</th>\n",
       "      <th>LUTnr</th>\n",
       "      <th>bitScore</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>tCount</th>\n",
       "      <th>mCount</th>\n",
       "      <th>Mode</th>\n",
       "      <th>Reads</th>\n",
       "      <th>identity</th>\n",
       "      <th>alignmentLength</th>\n",
       "      <th>gapOpens</th>\n",
       "      <th>q_start</th>\n",
       "      <th>q_end</th>\n",
       "      <th>s_start</th>\n",
       "      <th>s_end</th>\n",
       "      <th>evalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>28199</th>\n",
       "      <td>TCGCCGCGAGTACGTACGCATGGCCGT</td>\n",
       "      <td>seq_22420</td>\n",
       "      <td>44.525</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10883</th>\n",
       "      <td>TCGACGTGATTGAGTATGCAAGCCTGT</td>\n",
       "      <td>seq_22420</td>\n",
       "      <td>47.300</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31169</th>\n",
       "      <td>TCGCCTTCAACACGTATACAAATGCGC</td>\n",
       "      <td>seq_22420</td>\n",
       "      <td>47.300</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3027</th>\n",
       "      <td>TCGAAATCTACGAAGATACCCATAATT</td>\n",
       "      <td>seq_22420</td>\n",
       "      <td>47.300</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5469</th>\n",
       "      <td>TCGAATGAAAGATACAAATGAACAATT</td>\n",
       "      <td>seq_22420</td>\n",
       "      <td>47.300</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4101608</th>\n",
       "      <td>TCGGCGTGATGATATGCTCCAATGAAC</td>\n",
       "      <td>seq_22420</td>\n",
       "      <td>47.300</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Amb</td>\n",
       "      <td>CTGTGCCCACCAGCTGCGCCGAGGC</td>\n",
       "      <td>100.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>1.030000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4082730</th>\n",
       "      <td>TCGCCGCACACAAACAAACCAACCCCA</td>\n",
       "      <td>seq_22420</td>\n",
       "      <td>30.700</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Amb</td>\n",
       "      <td>CTGTCCCCACCCGCTCCGCCGAGGC</td>\n",
       "      <td>88.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>1.000000e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4094553</th>\n",
       "      <td>TCGCATTATTTGTACGAACAAGCCAGC</td>\n",
       "      <td>seq_22420</td>\n",
       "      <td>47.300</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Amb</td>\n",
       "      <td>CTGTGCCCACCAGCTGCGCCGAGGC</td>\n",
       "      <td>100.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>1.030000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3992956</th>\n",
       "      <td>TCGATTTCAACCCGTCTTTCCAGCAAT</td>\n",
       "      <td>seq_22420</td>\n",
       "      <td>47.300</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Amb</td>\n",
       "      <td>CTGTGCCCACCAGCTGCGCCGAGGC</td>\n",
       "      <td>100.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>1.030000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4007772</th>\n",
       "      <td>TCGCATCCTTCCATGCCTGCCACACTA</td>\n",
       "      <td>seq_22420</td>\n",
       "      <td>28.800</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Amb</td>\n",
       "      <td>CTGGCCTGCGGCTCAGCTGGGGGGC</td>\n",
       "      <td>88.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>3.700000e-02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>184 rows  16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  BC      LUTnr  bitScore  mismatches  tCount  \\\n",
       "28199    TCGCCGCGAGTACGTACGCATGGCCGT  seq_22420    44.525         0.0       4   \n",
       "10883    TCGACGTGATTGAGTATGCAAGCCTGT  seq_22420    47.300         0.0       3   \n",
       "31169    TCGCCTTCAACACGTATACAAATGCGC  seq_22420    47.300         0.0       2   \n",
       "3027     TCGAAATCTACGAAGATACCCATAATT  seq_22420    47.300         0.0       2   \n",
       "5469     TCGAATGAAAGATACAAATGAACAATT  seq_22420    47.300         0.0       2   \n",
       "...                              ...        ...       ...         ...     ...   \n",
       "4101608  TCGGCGTGATGATATGCTCCAATGAAC  seq_22420    47.300         0.0       1   \n",
       "4082730  TCGCCGCACACAAACAAACCAACCCCA  seq_22420    30.700         3.0       1   \n",
       "4094553  TCGCATTATTTGTACGAACAAGCCAGC  seq_22420    47.300         0.0       1   \n",
       "3992956  TCGATTTCAACCCGTCTTTCCAGCAAT  seq_22420    47.300         0.0       1   \n",
       "4007772  TCGCATCCTTCCATGCCTGCCACACTA  seq_22420    28.800         2.0       1   \n",
       "\n",
       "         mCount Mode                      Reads  identity  alignmentLength  \\\n",
       "28199         4  Def                        NaN       NaN              NaN   \n",
       "10883         3  Def                        NaN       NaN              NaN   \n",
       "31169         2  Def                        NaN       NaN              NaN   \n",
       "3027          2  Def                        NaN       NaN              NaN   \n",
       "5469          2  Def                        NaN       NaN              NaN   \n",
       "...         ...  ...                        ...       ...              ...   \n",
       "4101608       1  Amb  CTGTGCCCACCAGCTGCGCCGAGGC     100.0             25.0   \n",
       "4082730       1  Amb  CTGTCCCCACCCGCTCCGCCGAGGC      88.0             25.0   \n",
       "4094553       1  Amb  CTGTGCCCACCAGCTGCGCCGAGGC     100.0             25.0   \n",
       "3992956       1  Amb  CTGTGCCCACCAGCTGCGCCGAGGC     100.0             25.0   \n",
       "4007772       1  Amb  CTGGCCTGCGGCTCAGCTGGGGGGC      88.0             25.0   \n",
       "\n",
       "         gapOpens  q_start  q_end  s_start  s_end        evalue  \n",
       "28199         NaN      NaN    NaN      NaN    NaN           NaN  \n",
       "10883         NaN      NaN    NaN      NaN    NaN           NaN  \n",
       "31169         NaN      NaN    NaN      NaN    NaN           NaN  \n",
       "3027          NaN      NaN    NaN      NaN    NaN           NaN  \n",
       "5469          NaN      NaN    NaN      NaN    NaN           NaN  \n",
       "...           ...      ...    ...      ...    ...           ...  \n",
       "4101608       0.0      1.0   25.0     20.0   44.0  1.030000e-07  \n",
       "4082730       0.0      1.0   25.0     20.0   44.0  1.000000e-02  \n",
       "4094553       0.0      1.0   25.0     20.0   44.0  1.030000e-07  \n",
       "3992956       0.0      1.0   25.0     20.0   44.0  1.030000e-07  \n",
       "4007772       1.0      1.0   25.0     47.0   24.0  3.700000e-02  \n",
       "\n",
       "[184 rows x 16 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get all entries where mcount == total tcount\n",
    "new = p005[p005['mCount'] == p005['tCount']].sort_values(by='mCount', ascending=False)\n",
    "new[new[\"LUTnr\"] == \"seq_22420\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group</th>\n",
       "      <th>Category</th>\n",
       "      <th>AAstart</th>\n",
       "      <th>AAend</th>\n",
       "      <th>Structure</th>\n",
       "      <th>Peptide</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>seqlength</th>\n",
       "      <th>...</th>\n",
       "      <th>LUTnr</th>\n",
       "      <th>RNAcount</th>\n",
       "      <th>Normalized_RNAcount</th>\n",
       "      <th>bitScore</th>\n",
       "      <th>BC_count</th>\n",
       "      <th>BC_adjusted_count</th>\n",
       "      <th>AAwidth</th>\n",
       "      <th>AAseqlength</th>\n",
       "      <th>AA_pos</th>\n",
       "      <th>AA_rel_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DNA_AAVlib_DNAse_30cpc_1</td>\n",
       "      <td>Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>14aa</td>\n",
       "      <td>AADGYLPDWLEDNL</td>\n",
       "      <td>3</td>\n",
       "      <td>45</td>\n",
       "      <td>42</td>\n",
       "      <td>2208</td>\n",
       "      <td>...</td>\n",
       "      <td>seq_127</td>\n",
       "      <td>251</td>\n",
       "      <td>1050.197240</td>\n",
       "      <td>79.800820</td>\n",
       "      <td>1</td>\n",
       "      <td>1050.197240</td>\n",
       "      <td>14</td>\n",
       "      <td>736</td>\n",
       "      <td>8</td>\n",
       "      <td>0.010870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DNA_AAVlib_DNAse_30cpc_1</td>\n",
       "      <td>Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>14aa</td>\n",
       "      <td>ADGYLPDWLEDNLS</td>\n",
       "      <td>6</td>\n",
       "      <td>48</td>\n",
       "      <td>42</td>\n",
       "      <td>2208</td>\n",
       "      <td>...</td>\n",
       "      <td>seq_231</td>\n",
       "      <td>50</td>\n",
       "      <td>209.202637</td>\n",
       "      <td>80.895238</td>\n",
       "      <td>1</td>\n",
       "      <td>209.202637</td>\n",
       "      <td>14</td>\n",
       "      <td>736</td>\n",
       "      <td>9</td>\n",
       "      <td>0.012228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNA_AAVlib_DNAse_30cpc_1</td>\n",
       "      <td>Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>14aa</td>\n",
       "      <td>DGYLPDWLEDNLSE</td>\n",
       "      <td>9</td>\n",
       "      <td>51</td>\n",
       "      <td>42</td>\n",
       "      <td>2208</td>\n",
       "      <td>...</td>\n",
       "      <td>seq_324</td>\n",
       "      <td>48</td>\n",
       "      <td>200.834532</td>\n",
       "      <td>86.100000</td>\n",
       "      <td>1</td>\n",
       "      <td>200.834532</td>\n",
       "      <td>14</td>\n",
       "      <td>736</td>\n",
       "      <td>10</td>\n",
       "      <td>0.013587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DNA_AAVlib_DNAse_30cpc_1</td>\n",
       "      <td>Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>14aaA5</td>\n",
       "      <td>DGYLPDWLEDNLSE</td>\n",
       "      <td>9</td>\n",
       "      <td>51</td>\n",
       "      <td>42</td>\n",
       "      <td>2208</td>\n",
       "      <td>...</td>\n",
       "      <td>seq_44928</td>\n",
       "      <td>11</td>\n",
       "      <td>46.024580</td>\n",
       "      <td>106.733333</td>\n",
       "      <td>1</td>\n",
       "      <td>46.024580</td>\n",
       "      <td>14</td>\n",
       "      <td>736</td>\n",
       "      <td>10</td>\n",
       "      <td>0.013587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNA_AAVlib_DNAse_30cpc_1</td>\n",
       "      <td>Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1</td>\n",
       "      <td>3</td>\n",
       "      <td>25</td>\n",
       "      <td>22aa</td>\n",
       "      <td>DGYLPDWLEDNLSEGIREWWDL</td>\n",
       "      <td>9</td>\n",
       "      <td>75</td>\n",
       "      <td>66</td>\n",
       "      <td>2208</td>\n",
       "      <td>...</td>\n",
       "      <td>seq_76600</td>\n",
       "      <td>53</td>\n",
       "      <td>221.754796</td>\n",
       "      <td>88.673529</td>\n",
       "      <td>1</td>\n",
       "      <td>221.754796</td>\n",
       "      <td>22</td>\n",
       "      <td>736</td>\n",
       "      <td>14</td>\n",
       "      <td>0.019022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Group                                        Category  \\\n",
       "0  DNA_AAVlib_DNAse_30cpc_1  Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1   \n",
       "1  DNA_AAVlib_DNAse_30cpc_1  Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1   \n",
       "2  DNA_AAVlib_DNAse_30cpc_1  Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1   \n",
       "3  DNA_AAVlib_DNAse_30cpc_1  Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1   \n",
       "4  DNA_AAVlib_DNAse_30cpc_1  Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1   \n",
       "\n",
       "   AAstart  AAend Structure                 Peptide  start  end  width  \\\n",
       "0        1     15      14aa          AADGYLPDWLEDNL      3   45     42   \n",
       "1        2     16      14aa          ADGYLPDWLEDNLS      6   48     42   \n",
       "2        3     17      14aa          DGYLPDWLEDNLSE      9   51     42   \n",
       "3        3     17    14aaA5          DGYLPDWLEDNLSE      9   51     42   \n",
       "4        3     25      22aa  DGYLPDWLEDNLSEGIREWWDL      9   75     66   \n",
       "\n",
       "   seqlength  ...      LUTnr  RNAcount  Normalized_RNAcount    bitScore  \\\n",
       "0       2208  ...    seq_127       251          1050.197240   79.800820   \n",
       "1       2208  ...    seq_231        50           209.202637   80.895238   \n",
       "2       2208  ...    seq_324        48           200.834532   86.100000   \n",
       "3       2208  ...  seq_44928        11            46.024580  106.733333   \n",
       "4       2208  ...  seq_76600        53           221.754796   88.673529   \n",
       "\n",
       "  BC_count BC_adjusted_count  AAwidth  AAseqlength  AA_pos  AA_rel_pos  \n",
       "0        1       1050.197240       14          736       8    0.010870  \n",
       "1        1        209.202637       14          736       9    0.012228  \n",
       "2        1        200.834532       14          736      10    0.013587  \n",
       "3        1         46.024580       14          736      10    0.013587  \n",
       "4        1        221.754796       22          736      14    0.019022  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv('/lunarc/nobackup/projects/lu2024-17-19/J_rAAven/BRAVE/Brave/final_fragments_summary.csv')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"BC_count\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group</th>\n",
       "      <th>Category</th>\n",
       "      <th>AAstart</th>\n",
       "      <th>AAend</th>\n",
       "      <th>Structure</th>\n",
       "      <th>Peptide</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>seqlength</th>\n",
       "      <th>...</th>\n",
       "      <th>LUTnr</th>\n",
       "      <th>RNAcount</th>\n",
       "      <th>Normalized_RNAcount</th>\n",
       "      <th>bitScore</th>\n",
       "      <th>BC_count</th>\n",
       "      <th>BC_adjusted_count</th>\n",
       "      <th>AAwidth</th>\n",
       "      <th>AAseqlength</th>\n",
       "      <th>AA_pos</th>\n",
       "      <th>AA_rel_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DNA_AAVlib_DNAse_30cpc_1</td>\n",
       "      <td>Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>14aa</td>\n",
       "      <td>AADGYLPDWLEDNL</td>\n",
       "      <td>3</td>\n",
       "      <td>45</td>\n",
       "      <td>42</td>\n",
       "      <td>2208</td>\n",
       "      <td>...</td>\n",
       "      <td>seq_127</td>\n",
       "      <td>251</td>\n",
       "      <td>1050.197240</td>\n",
       "      <td>79.800820</td>\n",
       "      <td>12</td>\n",
       "      <td>1050.197240</td>\n",
       "      <td>14</td>\n",
       "      <td>736</td>\n",
       "      <td>8</td>\n",
       "      <td>0.010870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DNA_AAVlib_DNAse_30cpc_1</td>\n",
       "      <td>Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>14aa</td>\n",
       "      <td>ADGYLPDWLEDNLS</td>\n",
       "      <td>6</td>\n",
       "      <td>48</td>\n",
       "      <td>42</td>\n",
       "      <td>2208</td>\n",
       "      <td>...</td>\n",
       "      <td>seq_231</td>\n",
       "      <td>50</td>\n",
       "      <td>209.202637</td>\n",
       "      <td>80.895238</td>\n",
       "      <td>3</td>\n",
       "      <td>209.202637</td>\n",
       "      <td>14</td>\n",
       "      <td>736</td>\n",
       "      <td>9</td>\n",
       "      <td>0.012228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNA_AAVlib_DNAse_30cpc_1</td>\n",
       "      <td>Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>14aa</td>\n",
       "      <td>DGYLPDWLEDNLSE</td>\n",
       "      <td>9</td>\n",
       "      <td>51</td>\n",
       "      <td>42</td>\n",
       "      <td>2208</td>\n",
       "      <td>...</td>\n",
       "      <td>seq_324</td>\n",
       "      <td>48</td>\n",
       "      <td>200.834532</td>\n",
       "      <td>86.100000</td>\n",
       "      <td>5</td>\n",
       "      <td>200.834532</td>\n",
       "      <td>14</td>\n",
       "      <td>736</td>\n",
       "      <td>10</td>\n",
       "      <td>0.013587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DNA_AAVlib_DNAse_30cpc_1</td>\n",
       "      <td>Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>14aaA5</td>\n",
       "      <td>DGYLPDWLEDNLSE</td>\n",
       "      <td>9</td>\n",
       "      <td>51</td>\n",
       "      <td>42</td>\n",
       "      <td>2208</td>\n",
       "      <td>...</td>\n",
       "      <td>seq_44928</td>\n",
       "      <td>11</td>\n",
       "      <td>46.024580</td>\n",
       "      <td>106.733333</td>\n",
       "      <td>2</td>\n",
       "      <td>46.024580</td>\n",
       "      <td>14</td>\n",
       "      <td>736</td>\n",
       "      <td>10</td>\n",
       "      <td>0.013587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNA_AAVlib_DNAse_30cpc_1</td>\n",
       "      <td>Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1</td>\n",
       "      <td>3</td>\n",
       "      <td>25</td>\n",
       "      <td>22aa</td>\n",
       "      <td>DGYLPDWLEDNLSEGIREWWDL</td>\n",
       "      <td>9</td>\n",
       "      <td>75</td>\n",
       "      <td>66</td>\n",
       "      <td>2208</td>\n",
       "      <td>...</td>\n",
       "      <td>seq_76600</td>\n",
       "      <td>53</td>\n",
       "      <td>221.754796</td>\n",
       "      <td>88.673529</td>\n",
       "      <td>3</td>\n",
       "      <td>221.754796</td>\n",
       "      <td>22</td>\n",
       "      <td>736</td>\n",
       "      <td>14</td>\n",
       "      <td>0.019022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Group                                        Category  \\\n",
       "0  DNA_AAVlib_DNAse_30cpc_1  Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1   \n",
       "1  DNA_AAVlib_DNAse_30cpc_1  Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1   \n",
       "2  DNA_AAVlib_DNAse_30cpc_1  Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1   \n",
       "3  DNA_AAVlib_DNAse_30cpc_1  Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1   \n",
       "4  DNA_AAVlib_DNAse_30cpc_1  Adeno-associated virus,VP1,AAV1,,#125,AAV1-VP1   \n",
       "\n",
       "   AAstart  AAend Structure                 Peptide  start  end  width  \\\n",
       "0        1     15      14aa          AADGYLPDWLEDNL      3   45     42   \n",
       "1        2     16      14aa          ADGYLPDWLEDNLS      6   48     42   \n",
       "2        3     17      14aa          DGYLPDWLEDNLSE      9   51     42   \n",
       "3        3     17    14aaA5          DGYLPDWLEDNLSE      9   51     42   \n",
       "4        3     25      22aa  DGYLPDWLEDNLSEGIREWWDL      9   75     66   \n",
       "\n",
       "   seqlength  ...      LUTnr  RNAcount  Normalized_RNAcount    bitScore  \\\n",
       "0       2208  ...    seq_127       251          1050.197240   79.800820   \n",
       "1       2208  ...    seq_231        50           209.202637   80.895238   \n",
       "2       2208  ...    seq_324        48           200.834532   86.100000   \n",
       "3       2208  ...  seq_44928        11            46.024580  106.733333   \n",
       "4       2208  ...  seq_76600        53           221.754796   88.673529   \n",
       "\n",
       "  BC_count BC_adjusted_count  AAwidth  AAseqlength  AA_pos  AA_rel_pos  \n",
       "0       12       1050.197240       14          736       8    0.010870  \n",
       "1        3        209.202637       14          736       9    0.012228  \n",
       "2        5        200.834532       14          736      10    0.013587  \n",
       "3        2         46.024580       14          736      10    0.013587  \n",
       "4        3        221.754796       22          736      14    0.019022  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"BC_count\"] = df['BC'].str.split(',').apply(len)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library_barcodes.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Category', 'GeneName', 'AAstart', 'AAend', 'Structure', 'Peptide',\n",
       "       'Sequence', 'LUTnr', 'start', 'end', 'width', 'BC', 'bitScore',\n",
       "       'mismatches', 'tCount', 'mCount', 'Mode'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "library_barcodes.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library_barcodes = library_barcodes.drop(columns=[ 'Reads', 'identity','alignmentLength', 'gapOpens', 'q_start', 'q_end', 's_start', 's_end','evalue'])\n",
    "library_barcodes.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reorganize columns\n",
    "library_barcodes = library_barcodes[['Category', 'GeneName', 'LUTnr',  'BC', 'AAstart', 'AAend', 'Structure', 'Peptide',\n",
    "    'start', 'end', 'width', 'Sequence', 'bitScore',\n",
    "    'mismatches', 'tCount', 'mCount', 'RNAcount', 'Mode']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Category</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>LUTnr</th>\n",
       "      <th>BC</th>\n",
       "      <th>AAstart</th>\n",
       "      <th>AAend</th>\n",
       "      <th>Structure</th>\n",
       "      <th>Peptide</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>bitScore</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>tCount</th>\n",
       "      <th>mCount</th>\n",
       "      <th>Mode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Neurotrophin</td>\n",
       "      <td>NGF</td>\n",
       "      <td>seq_1</td>\n",
       "      <td>CATGCTATAAGGGAAGATTT</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>14aa</td>\n",
       "      <td>EPYTDSNVPEGDSV</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG</td>\n",
       "      <td>114.00</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Def</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Neurotrophin</td>\n",
       "      <td>NGF</td>\n",
       "      <td>seq_1</td>\n",
       "      <td>GTGTGTTGCCTTCATTCAGC</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>14aa</td>\n",
       "      <td>EPYTDSNVPEGDSV</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG</td>\n",
       "      <td>117.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Def</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Neurotrophin</td>\n",
       "      <td>NGF</td>\n",
       "      <td>seq_1</td>\n",
       "      <td>ACATGTTCATGTGCGCGCTG</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>14aa</td>\n",
       "      <td>EPYTDSNVPEGDSV</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG</td>\n",
       "      <td>75.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>Def</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Neurotrophin</td>\n",
       "      <td>NGF</td>\n",
       "      <td>seq_1</td>\n",
       "      <td>ATAGCAGGATGTCAGTGTGG</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>14aa</td>\n",
       "      <td>EPYTDSNVPEGDSV</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG</td>\n",
       "      <td>125.25</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>Def</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Neurotrophin</td>\n",
       "      <td>NGF</td>\n",
       "      <td>seq_1</td>\n",
       "      <td>ATGTGTTTGAAGGATGGATT</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>14aa</td>\n",
       "      <td>EPYTDSNVPEGDSV</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG</td>\n",
       "      <td>126.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>Def</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Category GeneName  LUTnr                    BC  AAstart  AAend  \\\n",
       "0  Neurotrophin      NGF  seq_1  CATGCTATAAGGGAAGATTT        0     14   \n",
       "1  Neurotrophin      NGF  seq_1  GTGTGTTGCCTTCATTCAGC        0     14   \n",
       "2  Neurotrophin      NGF  seq_1  ACATGTTCATGTGCGCGCTG        0     14   \n",
       "3  Neurotrophin      NGF  seq_1  ATAGCAGGATGTCAGTGTGG        0     14   \n",
       "4  Neurotrophin      NGF  seq_1  ATGTGTTTGAAGGATGGATT        0     14   \n",
       "\n",
       "  Structure         Peptide  start  end  width  \\\n",
       "0      14aa  EPYTDSNVPEGDSV      0   42     42   \n",
       "1      14aa  EPYTDSNVPEGDSV      0   42     42   \n",
       "2      14aa  EPYTDSNVPEGDSV      0   42     42   \n",
       "3      14aa  EPYTDSNVPEGDSV      0   42     42   \n",
       "4      14aa  EPYTDSNVPEGDSV      0   42     42   \n",
       "\n",
       "                                     Sequence  bitScore  mismatches  tCount  \\\n",
       "0  GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG    114.00         2.5       2   \n",
       "1  GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG    117.00         1.0       2   \n",
       "2  GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG     75.00         0.0       5   \n",
       "3  GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG    125.25         0.0       5   \n",
       "4  GAGCCCTACACCGACAGCAACGTGCCCGAGGGCGACAGCGTG    126.00         1.0       7   \n",
       "\n",
       "   mCount Mode  \n",
       "0       2  Def  \n",
       "1       2  Def  \n",
       "2       3  Def  \n",
       "3       4  Def  \n",
       "4       3  Def  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "library_barcodes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3261305, 18)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "library_barcodes = pd.read_csv('/lunarc/nobackup/projects/lu2024-17-19/J_rAAven/BRAVE/0_data/pos_library_barcodes.csv')\n",
    "library_barcodes.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BC</th>\n",
       "      <th>LUTnr</th>\n",
       "      <th>bitScore</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>tCount</th>\n",
       "      <th>mCount</th>\n",
       "      <th>Mode</th>\n",
       "      <th>Reads</th>\n",
       "      <th>identity</th>\n",
       "      <th>alignmentLength</th>\n",
       "      <th>...</th>\n",
       "      <th>Category</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>AAstart</th>\n",
       "      <th>AAend</th>\n",
       "      <th>Structure</th>\n",
       "      <th>Peptide</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTGCCTAGACTCCCAGCCTT</td>\n",
       "      <td>seq_88507</td>\n",
       "      <td>122.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Amb</td>\n",
       "      <td>CTTACGCCACCGGCAGCAGCCCCGACGTGCAGATCACCGCCAGCCT...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Theilers murine encephalomyelitis virus</td>\n",
       "      <td>TMEV-VP2</td>\n",
       "      <td>234</td>\n",
       "      <td>256</td>\n",
       "      <td>22aa</td>\n",
       "      <td>YATGSSPDVQITASLQPVNPVF</td>\n",
       "      <td>TACGCCACCGGCAGCAGCCCCGACGTGCAGATCACCGCCAGCCTGC...</td>\n",
       "      <td>702</td>\n",
       "      <td>768</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTACGTGCACGCGTCATT</td>\n",
       "      <td>seq_78536</td>\n",
       "      <td>78.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Neurotoxin</td>\n",
       "      <td>BoNT/F Hc</td>\n",
       "      <td>204</td>\n",
       "      <td>226</td>\n",
       "      <td>22aa</td>\n",
       "      <td>KVFDTELGKTEIETLYSDEPDP</td>\n",
       "      <td>AAGGTGTTCGACACCGAGCTGGGCAAGACCGAGATCGAGACCCTGT...</td>\n",
       "      <td>612</td>\n",
       "      <td>678</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTACGGACATCAGCATGC</td>\n",
       "      <td>seq_34770</td>\n",
       "      <td>76.200000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Human immunodeficiency virus type 1</td>\n",
       "      <td>HIV-gp120-TH90129_1</td>\n",
       "      <td>210</td>\n",
       "      <td>224</td>\n",
       "      <td>14aa</td>\n",
       "      <td>GPCTNVSTVQCTHG</td>\n",
       "      <td>GGCCCCTGCACCAACGTGAGCACCGTGCAGTGCACCCACGGC</td>\n",
       "      <td>630</td>\n",
       "      <td>672</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTGTGTATGGGCGTATGG</td>\n",
       "      <td>seq_21642</td>\n",
       "      <td>78.700000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Herpes simplex virus type 1</td>\n",
       "      <td>HSV-1-pUS8</td>\n",
       "      <td>393</td>\n",
       "      <td>407</td>\n",
       "      <td>14aa</td>\n",
       "      <td>GADLAEPTHPHVGA</td>\n",
       "      <td>GGCGCCGACCTGGCCGAGCCCACCCACCCCCACGTGGGCGCC</td>\n",
       "      <td>1179</td>\n",
       "      <td>1221</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GTACGCGTAGGAGTCCAC</td>\n",
       "      <td>seq_22869</td>\n",
       "      <td>78.700000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>Def</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Herpes simplex virus type 1</td>\n",
       "      <td>HSV-1-pUL19</td>\n",
       "      <td>1112</td>\n",
       "      <td>1126</td>\n",
       "      <td>14aa</td>\n",
       "      <td>ATVRNPVTDMGNLP</td>\n",
       "      <td>GCCACCGTGCGGAACCCCGTGACCGACATGGGCAACCTGCCC</td>\n",
       "      <td>3336</td>\n",
       "      <td>3378</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     BC      LUTnr    bitScore  mismatches  tCount  mCount  \\\n",
       "0  GTGCCTAGACTCCCAGCCTT  seq_88507  122.000000         0.0       1       1   \n",
       "1    GTACGTGCACGCGTCATT  seq_78536   78.666667         0.0       3       3   \n",
       "2    GTACGGACATCAGCATGC  seq_34770   76.200000         0.0       5       3   \n",
       "3    GTGTGTATGGGCGTATGG  seq_21642   78.700000         0.0       2       2   \n",
       "4    GTACGCGTAGGAGTCCAC  seq_22869   78.700000         0.0       2       1   \n",
       "\n",
       "  Mode                                              Reads  identity  \\\n",
       "0  Amb  CTTACGCCACCGGCAGCAGCCCCGACGTGCAGATCACCGCCAGCCT...     100.0   \n",
       "1  Def                                                NaN       NaN   \n",
       "2  Def                                                NaN       NaN   \n",
       "3  Def                                                NaN       NaN   \n",
       "4  Def                                                NaN       NaN   \n",
       "\n",
       "   alignmentLength  ...                                 Category  \\\n",
       "0             66.0  ...  Theilers murine encephalomyelitis virus   \n",
       "1              NaN  ...                               Neurotoxin   \n",
       "2              NaN  ...      Human immunodeficiency virus type 1   \n",
       "3              NaN  ...              Herpes simplex virus type 1   \n",
       "4              NaN  ...              Herpes simplex virus type 1   \n",
       "\n",
       "              GeneName  AAstart  AAend  Structure                 Peptide  \\\n",
       "0             TMEV-VP2      234    256       22aa  YATGSSPDVQITASLQPVNPVF   \n",
       "1            BoNT/F Hc      204    226       22aa  KVFDTELGKTEIETLYSDEPDP   \n",
       "2  HIV-gp120-TH90129_1      210    224       14aa          GPCTNVSTVQCTHG   \n",
       "3           HSV-1-pUS8      393    407       14aa          GADLAEPTHPHVGA   \n",
       "4          HSV-1-pUL19     1112   1126       14aa          ATVRNPVTDMGNLP   \n",
       "\n",
       "                                            Sequence start   end  width  \n",
       "0  TACGCCACCGGCAGCAGCCCCGACGTGCAGATCACCGCCAGCCTGC...   702   768     66  \n",
       "1  AAGGTGTTCGACACCGAGCTGGGCAAGACCGAGATCGAGACCCTGT...   612   678     66  \n",
       "2         GGCCCCTGCACCAACGTGAGCACCGTGCAGTGCACCCACGGC   630   672     42  \n",
       "3         GGCGCCGACCTGGCCGAGCCCACCCACCCCCACGTGGGCGCC  1179  1221     42  \n",
       "4         GCCACCGTGCGGAACCCCGTGACCGACATGGGCAACCTGCCC  3336  3378     42  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "found_barcodes = pd.read_csv('/lunarc/nobackup/projects/lu2024-17-19/J_rAAven/BRAVE/0_data/found_barcodes/found.mRNA_30cpc_Th.csv')\n",
    "found_barcodes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# concat dataframes but only keep the columns that are in both dataframes\n",
    "combined = pd.concat([library_barcodes, found_barcodes], join='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ref_sequence_length_df(file_path:str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Reads a FASTA file containing reference sequences and returns a DataFrame with sequence ids and lengths.\n",
    "    Args:\n",
    "        file_path (str): path to the file containing the reference sequences\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: DataFrame with the reference sequence ids and lengths\n",
    "    \"\"\"\n",
    "    \n",
    "    # read in file with original sequences\n",
    "    seqs_original = list(SeqIO.parse(file_path, \"fasta\"))\n",
    "    # translate sequences to amino acids\n",
    "    seq = [str(seq_record.seq).strip() for seq_record in seqs_original]\n",
    "    # get sequence ids\n",
    "    ids = [seq_record.description for seq_record in seqs_original]\n",
    "    # get sequence categories\n",
    "    categories = [id.split(\",\")[0] for id in ids]\n",
    "    # get the GeneName\n",
    "    gene_names = [id.split(\",\")[5] for id in ids]\n",
    "    \n",
    "    # create a DataFrame with sreference sequence ids and lengths\n",
    "    ref_seq_len_df = pd.DataFrame({\"Category\": categories,\n",
    "                                    \"GeneName\": gene_names,\n",
    "                                    \"seqlength\": [len(s) for s in seq]})\n",
    "    \n",
    "    return ref_seq_len_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = get_ref_sequence_length_df('/lunarc/nobackup/projects/lu2024-17-19/J_rAAven/BRAVE/input/DNA-lib_RetrogradeTransport.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA_pscAAVlib_Prep2_1\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "test = \"DNA_pscAAVlib_Prep2_1.fastq.gz\"\n",
    "test_name=os.path.basename(test).replace('.fastq.gz', '')\n",
    "print(test_name)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
